gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspSubplugin.kt

```diff
@@ -136,6 +137,13 @@ class KspGradleSubplugin @Inject internal constructor(private val registry: Tool
             kspExtension.apOptions.forEach {
                 options += SubpluginOption("apoption", "${it.key}=${it.value}")
             }
+            kspExtension.commandLineArgumentProviders.forEach {
+                val argument = it.asArguments().joinToString("")
+                if (!argument.matches(Regex("/S=/S"))) {
+                    throw IllegalArgumentException("KSP apoption does not match /S=/S: $argument")
+                }
+                options += SubpluginOption("apoption", argument)
```

Comment:
This differs from [the kapt implementation](https://github.com/JetBrains/kotlin/blob/68f4580492894ee18a60319547dc376983a764d3/libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/kapt/Kapt3KotlinGradleSubplugin.kt#L392-L405) in a couple of ways:

1️⃣ A `SubpluginOption`'s value - which contains an absolute path in the case of `room.schemaLocation` - [becomes a task input](https://github.com/JetBrains/kotlin/blob/8753ea592f5c682bb94e0d24875113290512324a/libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/plugin/KotlinPlugin.kt#L1087-L1112). It might be worth adding an integration test that uses the local build cache from a different location and checking that there are cache hits?

2️⃣ A `CommandLineArgumentProvider` can have arbitrary lazy properties that are not necessarily safe to read at this point in the configuration phase. It'll appear to work for simple cases like `room.schemaLocation` pointing to a static directory, but I think it'll break when there are properties that can only be safely evaluated at execution time - see [this kapt bug](https://youtrack.jetbrains.com/issue/KT-39715).

For some reason `InternalSubpluginOption` doesn't support lazy values, so I think the workaround here would be to use `CompositeSubpluginOption("apoption", lazy { it.asArguments().joinToString("")) }, emptyList())` (you could also pass a list containing a single `InternalSubpluginOption` as the final argument, which probably makes more sense semantically but behaves equivalently in practice).

Reply:
Thanks for pointing this out. I will update to make this more in line with the KAPT approach.

Reply:
I think we need to take a step back and think about how this would work in concert with #861 before moving forward here.

Reply:
KSP didn't call `Task.registerSubpluginOptionsAsInputs`. It registers AP options as task inputs through [KspTask.apOptions](https://github.com/google/ksp/blob/510130e9d32623d0fe1f8498a38afc418f4ba6c3/gradle-plugin/src/main/kotlin/com/google/devtools/ksp/gradle/KspSubplugin.kt#L412) instead. The options are evaluated at execution time (wrapped by a provider in configuration time, and `get()` at execution time). Therefore the issue is slightly different: they need to be registered manually.

@runningcode We plan to build 1.0.5 release on Apr. 6. Can we land this first and handle variants in another patch?

Reply:
Thanks for the update.
Can you tell me which of the concerns that Phil raised should be addressed in this PR and how they should be addressed?

Reply:
Calling `asArguments()` and creating `SubpluginOption` here is fine because it is evaluated at execution time.

Reply:
To be clear, the only concern I have is that this may need to be wrapped by a provider:
```
kspTask.commandLineArgumentProviders.addAll(kspExtension.commandLineArgumentProviders)
```

Reply:
Ah, my mistake! I didn't check the call chain properly and assumed all `SubpluginOption`s went through that method.

I don't _think_ there's any need for further laziness, since calling `asArguments()` is already guarded by the provider that creates the list of options?